lib/lists.nix

```
@@ -637,6 +637,76 @@ rec {
        else if start + count > len then len - start
        else count);
 
+  /* The length of the common prefix of two lists.
+
+  Type: commonPrefixLength :: [a] -> [a] -> Int
+
+  Example:
+    commonPrefixLength [ 1 2 3 4 5 6 ] [ 1 2 4 8 ]
+    => 2
+    commonPrefixLength [ 1 2 3 ] [ 1 2 3 4 5 ]
+    => 3
+    commonPrefixLength [ 1 2 3 ] [ 4 5 6 ]
+    => 0
+  */
+  commonPrefixLength =
+    let
+      shorterFirst =
+        shorter: longer:
+        let
+          # We fold over the shorter list, no need to check elements of the
+          # longer list that couldn't be matched by the shorter one
+
+          # index invariant:
+          # - if index < 0 then (- index - 1) is the first index that differs between the two lists
+          # - if index >= 0 then all elements below that index match and s == elemAt shorter index
+          result = lib.foldl' (index: s:
+            if index < 0 then
+              # There's already a differing element before, no need to check anything further
+              index
+            else
+              # All previous elements match, let's also check this one
+              if elemAt longer index == s then
+                # If it does match, increase the index to go to the next one
+                index + 1
+              else
+                # Elements don't match
+                # Set index such that the invariant is satisfied:
+                # (invariant) oldIndex == - newIndex - 1
+                # => newIndex == - oldIndex - 1
+                - index - 1
+          ) 0 shorter;
+        in
+        if result < 0 then
+          # Elements at index (- index - 1) differ, which is also the length of the common prefix
+          - result - 1
+        else
+          # We reached the end of the shorter list, indicating that all elements matched the longer one
+          result;
+    in
+    list1: list2:
+    if length list1 < length list2 then
+      shorterFirst list1 list2
```

```suggestion
      commonPrefixLength' list1 list2
```

The other name doesn't really say what it's about.